### ISSUE TITLE
SCN vs. sh for gradients

### ISSUE THREAD

**Issue by doctormo** [jpakkane/capypdf#43]
There seems to be two ways to describe a gradient in PDF, either the PDF 1.2 SCN/scn method that sets a Pattern Type 2 as a gradient, or the sh command that directly links to a shading object.

The big difference is that SCN/scn can be used in strokes and fill, while sh is only available for fill, making it much more limited.

Do you know which capypdf is currently supporting?

**Comment by jpakkane** [comment]
`sh` seems to be the only supported one. I have not done proper `SCN/scn` support, probably because I did not have pattern support available at all at the time. That needs to be added.

**Comment by doctormo** [comment]
You can see the kind of construction from this cairo output:  [test-sf-grad.pdf](https://github.com/user-attachments/files/17806696/test-sf-grad.pdf)

So constructing a pattern from a shading object is going to be pretty critical, because I'll also need to construct soft masks for transparencies and trying to layer a individual transparency groups with soft masks is going to be a real PITA. But if I can SCN/scn the gradient, then the transparency will just mirror that same structure without extra groups.

**Comment by jpakkane** [comment]
Yes, that is definitely a thing that should be supported. However I will most likely be unable to work on that until next week.

**Comment by doctormo** [comment]
Looking at the code, I can see patterns are handled differently from other objects and are constructed in-stream from the given draw context.

If the same code pattern is used we might end up with:

```
rvoe<CapyPDF_PatternId> PdfDocument::add_pattern(PdfShading sh, double a, double b, double c, double d, double e, double f) {
  std::string shading_commands = serialise_shading(sh);
      auto pattern_dict = std::format(R"(<<
    /Type /Pattern
    /PatternType 2
    /Matrix {:f} {:f} {:f} {:f} {:f} {:f}
>>
)", a, b, c, d, e, f);

  return CapyPDF_PatternId{
          add_object(FullPDFObject{std::move(pattern_dict), std::string(shading_commands)})};
}
```

The refactoring needed, would be:

 1. Making `serialise_shading` do the switch between each of the shading type variants
 2. Make it return the shading dictionary without adding it as a Gfx reource object
 3. Refactor `add_shading` so it just calls `serialise_shading` and adds the shading object itself.

Before I do any of this work, I'd like your input @jpakkane

**Comment by doctormo** [comment]
I created an MR for the refactoring work so you can see what it looks like in situ

https://github.com/jpakkane/capypdf/pull/46/files

**Comment by doctormo** [comment]
This makes sense. I suspect you want to upgrade the Pattern to a full fat object so a matrix or other settings can be applied.

I feel like I've run this up to the line of what you find comfortable me contributing. So I'll leave this with you can work on something else for now.

I did work out how to do reflection and repeat gradient patterns though. So many things to test!
